### PR TITLE
Add flows list sc 16476

### PR DESCRIPTION
**Pull request by derek-globus**
Implements [sc-16476](https://app.shortcut.com/globus/story/16476/globus-cli-should-support-flow-list)

### Enhancements

* Added hidden flows list command
  * Usage: `globus flows list`

### Relevant Notes
1. The new `flows` subcommand is hidden (not listed in help) but still invokable as per [sc-16476](https://app.shortcut.com/globus/story/16476/globus-cli-should-support-flow-list) TAC.
2. The implementation here in standard cli output significantly truncates the fields returned from 27 down to 5
    * My rule of thumb was that list operations should provide the user enough info that they can intelligently select an object (flow in this case) to interact more with more through other apis.
      * The current list api simply returns the full flow db object including things like full definition, input schema, etc.
      * For a (mostly) full list of fields returned by the list api see [read-the-docs](https://globusonline.github.io/flows/#tag/Flows/paths/~1flows/get)
    * If the result is output as json instead of the standard table format, all fields returned by the api are included in this blob

### Testing
*Manual testing beyond the functional tests implemented*
No Args
```
(.venv) derekschlabach@Dereks-MacBook-Pro globus-cli % globus flows list
Flow ID                              | Title            | Owner            | Created At          | Updated At         
------------------------------------ | ---------------- | ---------------- | ------------------- | -------------------
f73633d2-9445-4558-8228-977f79d71971 | derek-dummy-flow | derek@globus.org | 2022-09-16 13:22:27 | 2022-09-19 16:42:19
```

View all flows I have access to
```
(.venv) derekschlabach@Dereks-MacBook-Pro globus-cli % globus flows list --filter-role flow_viewer
Flow ID                              | Title                                                   | Owner                                                        | Created At          | Updated At         
------------------------------------ | ------------------------------------------------------- | ------------------------------------------------------------ | ------------------- | -------------------
f73633d2-9445-4558-8228-977f79d71971 | derek-dummy-flow                                        | derek@globus.org                                             | 2022-09-16 13:22:27 | 2022-09-19 16:42:19
9b8c4c8d-ef05-462f-842b-6b7c3163d5b2 | Transfer specified Files v1                             | lei@globus.org                                               | 2022-05-09 16:54:44 | 2022-05-10 08:57:38
99791f7d-6c2c-4675-af4b-b927db68bad0 | Two Stage Globus Transfer                               | 8a0cdb6a-4ff4-4dcd-9ae4-6eb45611eef8@clients.auth.globus.org | 2022-03-30 10:24:57 | 2022-03-30 10:24:57
...truncated...
```

Further filter by text
```

(.venv) derekschlabach@Dereks-MacBook-Pro globus-cli % globus flows list --filter-role flow_viewer --filter-fulltext "Hello World"
Flow ID                              | Title                  | Owner                     | Created At          | Updated At         
------------------------------------ | ---------------------- | ------------------------- | ------------------- | -------------------
13b36c21-74e8-45f4-9b69-5f55b05a3366 | Hello, World (Example) | jbottigliero@uchicago.edu | 2021-10-18 14:19:35 | 2021-12-01 15:53:30
```

### CODE REVIEW

**Comment by sirosen** [comment]
I typically prefer to have things wait in the CLI if we're making upstream changes in SDK or elsewhere, so I think we're holding this for a little.

Even so, I've added myself and Aaron as reviewers and approved the workflow run. (We should get Derek added to the `globus` github org if we haven't already.) I'm looking forward to circling back on this when 1. I'm back at work 100% and 2. we have any upstream changes worked out.

**Comment by derek-globus** [comment]
Sounds fine, Stephen. Given that this is a hidden command still, I can't imagine there's significant urgency in getting it out beyond closing the open story.

I'll keep this pr open but hold off until the SDK `FlowsScopes` changes have been released then update this branch to get rid of the smells I introduced.

**Comment by derek-globus** [comment]
The sdk has been bumped and released to 3.12.0 to incorporate the `FlowsScopes.resource_server` bugfix. That change is now pulled into this pr.

File: tests/conftest.py
Comment (sirosen) [inline tests/conftest.py:83]: You may have noticed that the data above this is not uniform. The `TimerScopes` usage is the newest, and I think I'd like for us to follow that pattern (and go back to fix the other data here). If you agree that it's better, we'd get
`FlowsScopes.resource_server: _mock...(FlowsScopes.resource_server, FlowsScopes.view_flows)`.

Also note that this should match whatever we're requesting in `LoginManager`, if that should change.

**Review by sirosen — CHANGES_REQUESTED** [review]
This looks good! You've figured out how to wrangle several nontrivial parts of the CLI here (e.g. `LoginManager`), and most everything looks right and the tests look good.

There is one sticking point: the scopes. I think we need to act on that particular item before we can merge. Otherwise, we'll need to add support for `LoginManager` to inspect the scopes associated with the tokens it has in storage. I believe we have the data to do that, and it wouldn't necessarily be a bad internal feature to add, but I'd rather avoid making it a requirement at this point in time.

File: changelog.d/20220920_140606_derek_add_flows_list_sc_16476.md
Comment (derek-globus) [reply changelog.d/20220920_140606_derek_add_flows_list_sc_16476.md:5]: I hadn't thought about it this way before, but since you brought it up, I agree. I'll drop it

File: src/globus_cli/commands/flows/list.py
Comment (derek-globus) [reply src/globus_cli/commands/flows/list.py:44]: Oh neat, I didn't realize that `None` values would be treated the same as no keys for query params.

If that's true, then I agree 100% with you. Will update in new rev.

File: src/globus_cli/commands/flows/list.py
Comment (derek-globus) [reply src/globus_cli/commands/flows/list.py:16]: I generally prefer verbosity (assuming it's not too verbose) to falling back to defaults.

As much as possible (and as much as doesn't leave the code way too tiresome to read), I find explicit affirmations of default values (1) ensures I've actually thought about what the state should be, and (2) provides local context for any developers coming behind me who aren't familiar with click's specific defaulting stance to know more readily what the decorator is doing.

Not a hill I'd die on & one where I think is more often a judgement call (I don't for instance annotate every data class with `@dataclass(frozen=False)`) but in this case I'd like to keep the flag explicit unless it's something you do feel strongly about (in which case I'm happy changing it).

File: src/globus_cli/login_manager/manager.py
Comment (derek-globus) [reply src/globus_cli/login_manager/manager.py:65]: Yeah I'm on board with this.

When I was initially setting this up, I hadn't read through much of the auth-side of the code & thought it worked much closer to the dynamic scope authorizing approach so I was trying to limit to minimal scope.

I'll throw the whole gambit of scopes in there in the next rev.

File: tests/conftest.py
Comment (derek-globus) [reply tests/conftest.py:83]: I think I like the explicit string definitions here.

Generally speaking I'd prefer referencing resource server like is done in the login manager. But because this is a test not source, explicitly putting the resource server/expected scope strings here relies less on upstream definitions being correct (ie. the flows resource server) and actually catches more issues that could occur (like the fix we needed in sdk).

On the flip side, supposing these values changed in real life, the impact would just be a test suite breaks & is caught on automated test run before the cli goes out to customers. I'd argue the value of confirming our expectation of how things work is worth the cost of a very occasional dev process slowdown.

File: src/globus_cli/commands/flows/_common.py
Comment (derek-globus) [reply src/globus_cli/commands/flows/_common.py:9]: I like the field function syntax a lot.

It's very reminiscent of [dataclass fields](https://docs.python.org/3/library/dataclasses.html#dataclasses.field), so my natural inclination is to ask whether we can tap into something on that side to keep it closer to "any python dev could edit this" & remove repo-specific knowledge.

Some sort of declarative transformation(/validation?) layer probably exists through marshmallow or pydantic or something similar which taps into that exact syntactic style you've defined here. If not, I like your proposal #1 for explicit declarative transformation. Could support a `type` kwarg as well as a `transform` kwarg so if somebody really wants to go in and mess with the data in a non-repeatable way they have the freedom to do that.

Comment (sirosen) [reply src/globus_cli/commands/flows/_common.py:9]: Let me come back to this and see what I can do with dataclasses. That "feels right", and we may just need to leverage the `metadata` blob in the right way.

File: src/globus_cli/commands/flows/list.py
Comment (sirosen) [reply src/globus_cli/commands/flows/list.py:16]: My rationale for this is that we should never (IMO) be using `click.option(..., required=True)`. To me, `required=False` reads as redundant with `option` being the decorator name.

Like you, this is not some line in the sand which I refuse to cross, but I still advocate for following the existing style in the repo today.

File: tests/conftest.py
Comment (sirosen) [reply tests/conftest.py:89]: Yep, I'm convinced. 👍  I'll mark this resolved.

File: src/globus_cli/login_manager/manager.py
Comment (sirosen) [reply src/globus_cli/login_manager/manager.py:66]: This looks good to me for now. I'd love to see it reduce down to `FlowsScopes.all` in the future, but we'll leave that as a future refinement. (And we'll probably need to update some of the surrounding code to check the scopes on stored tokens.)

**Review by sirosen — COMMENTED** [review]
With the change to scopes, this looks good to me. I want a chance to test it myself before approving, and I would like to push slightly for the `required=False` change (rationale inline).

I'm going to mark some threads as resolved and we can follow up on the rest.

File: src/globus_cli/commands/flows/list.py
Comment (derek-globus) [reply src/globus_cli/commands/flows/list.py:16]: Sure, I'll drop the required in a new rev.

**Pull request merged**
